### ISSUE TITLE
Is it possible to grab weights from another operator using Transpiler?

### ISSUE THREAD

**Issue by wojtuss** [PaddlePaddle/Paddle#12507]
I'm working on fusing FC and GRU layers by the MKL-DNN GRU op for inference. Using the `inference_transpiler.py`, is it possible to grab the weights from the FC's MUL op and "add" it as a weight to the MKL-DNN-based GRU op?

The situation is as follows:
- in the GRUOpMaker there is a new input `WeightX` added (`AsDispensable()`),
- there is no `WeightX` input passed via `dynamic_gru` in `nn.py`, so the ordinary GRU could work,
- a new GRUMKLDNNKernel is created, which requires the `WeightX` tensor, it is called only when inferencing with MKL-DNN,
- in the `inference_transpiler.py` I am fusing the MUL+elementwise_add+GRU ops, but this requires grabbing the MUL's weights into the GRU's `WeightX` input.
How to achieve that? From the code of the transpiler, I am not really sure how to do that.

I know the "Pass" is comming to facilitate inference fusing, but until then I want to familiarize myself to and do the fusing using the transpiler.

**Comment by wojtuss** [comment]
@luotao1 , the simplest solution probably would be to insert a new GRU op with the `WeightX` parameter passed to it and to delete the FC and old GRU ops. However, is it possible to grab the FC's weights without deleting the GRU op?

**Comment by luotao1** [comment]
> I am fusing the MUL+elementwise_add+GRU ops, but this requires grabbing the MUL's weights into the GRU's WeightX input. the simplest solution probably would be to insert a new GRU op with the WeightX parameter passed to it and to delete the FC and old GRU ops.

Yes, this solution is OK, insert a new GRU op, and delete the FC and old GRU ops in program_desc.

> this requires grabbing the MUL's weights into the GRU's WeightX input.

You should do some weight computation and combination. For example, in conv+batch_norm, we directly add batch_norm weight into conv weight based on formula:
https://github.com/PaddlePaddle/Paddle/blob/3185d38af4eca081993845621943760fd254aa2b/python/paddle/fluid/transpiler/inference_transpiler.py#L115-L126
You can grab the MUL's weights into the GRU's WeightX input before the inference run.

> is it possible to grab the FC's weights without deleting the GRU op

I think it is possible. In conv+batch_norm, we don't remove the old conv, but only adjust the weight (including rename the weight name) instead.

**Comment by wojtuss** [comment]
@luotao1 , thanks!
I have started with the simplest approach (add a new GRU op, remove the old one and the FC). Also I modeled the fusion on the conv+batch_norm fusion. Although the transpiler seems to work I probably miss updating the connection between the input name `WeightX` and the actual data taken from the MUL op (say `fc_0.w_0_fuse_gru`) in the new GRU op, as the `WeightX` input comes empty to the MKL-DNN-based GRU kernel.
How can I fix the correspondence between the input name `WeightX` and the input arg name `fc_0.w_0_fuse_gru`?
